### ISSUE TITLE
E3-2: Implement experiment execution environment

### ISSUE THREAD

**Issue by jkbrooks** [jkbrooks/autoresearch_network#10]
## E3-2: Implement experiment execution environment

**Parent Epic:** #8 — Miner Implementation  
**Dependencies:** #9 (E3-1 — miner scaffold must exist)  
**Implementation order:** 2 of 9 (MVP)

---

### What This Is

An isolated execution environment that takes a `train.py` source string, writes it to a temp directory alongside the bundled `prepare.py`, runs it as a subprocess with timeout enforcement, captures all output, parses the metrics summary block, and returns a structured result. This is the engine that actually runs the 5-minute training experiments.

### Where It Goes

- `autoresearch/experiment_runner.py` — `ExperimentRunner` class + `RunResult` dataclass
- `autoresearch/data/prepare.py` — verbatim copy of https://github.com/karpathy/autoresearch/blob/master/prepare.py (MIT licensed)

### Technical Context

Each incoming validator request triggers one training run. The miner writes `train.py` to an isolated temp directory, copies/symlinks `prepare.py` alongside it, and runs `uv run train.py` as a subprocess. The training script runs for a fixed 5-minute wall-clock budget and prints a metrics summary block to stdout upon completion.

**The summary block** (printed by `train.py` lines 620–629):
```
---
val_bpb:          0.997900
training_seconds: 300.1
total_seconds:    325.9
peak_vram_mb:     45060.2
mfu_percent:      39.80
total_tokens_M:   499.6
num_steps:        953
num_params_M:     50.3
depth:            8
```

**Reference:** https://github.com/karpathy/autoresearch/blob/master/train.py#L620-L629

**Reference:** https://github.com/karpathy/autoresearch/blob/master/program.md — "Run the experiment: `uv run train.py > run.log 2>&1`" and timeout rules (kill if >10 min)

### Data Types

```python
from dataclasses import dataclass
from typing import Optional

@dataclass
class RunResult:
    val_bpb: Optional[float] = None
    training_seconds: Optional[float] = None
    total_seconds: Optional[float] = None
    peak_vram_mb: Optional[float] = None
    mfu_percent: Optional[float] = None
    total_tokens_m: Optional[float] = None
    num_steps: Optional[int] = None
    num_params_m: Optional[float] = None
    depth: Optional[int] = None
    run_log_tail: str = ""        # last 100 lines of stdout
    status: str = "pending"       # "success" | "crash" | "timeout"
```

### ExperimentRunner Interface

```python
class ExperimentRunner:
    def __init__(self, prepare_py_path: str = None):
        """
        prepare_py_path: path to bundled prepare.py.
        Defaults to autoresearch/data/prepare.py in the package.
        """

    def setup(self) -> bool:
        """
        One-time: run `uv run prepare.py` if ~/.cache/autoresearch/ doesn't have data shards.
        Returns True if data is ready, False on failure.
        Logs whether setup was run or skipped.
        """

    def run(self, train_py_source: str) -> RunResult:
        """
        1. Create temp directory
        2. Write train_py_source to temp/train.py
        3. Copy/symlink prepare.py to temp/prepare.py
        4. Symlink pyproject.toml if needed for `uv run` to work
        5. Run `uv run train.py > run.log 2>&1` with 600s timeout
        6. If timeout: kill subprocess tree, return RunResult(status="timeout")
        7. If non-zero exit: return RunResult(status="crash", run_log_tail=last 100 lines)
        8. Parse metrics from run.log
        9. Clean up temp directory
        10. Return RunResult(status="success", ...)
        """
```

### Parsing Logic

Use regex to extract each metric from the captured log:
```python
import re

def parse_metrics(log: str) -> dict:
    patterns = {
        "val_bpb": r"^val_bpb:\s+([\d.]+)",
        "training_seconds": r"^training_seconds:\s+([\d.]+)",
        "total_seconds": r"^total_seconds:\s+([\d.]+)",
        "peak_vram_mb": r"^peak_vram_mb:\s+([\d.]+)",
        "mfu_percent": r"^mfu_percent:\s+([\d.]+)",
        "total_tokens_M": r"^total_tokens_M:\s+([\d.]+)",
        "num_steps": r"^num_steps:\s+(\d+)",
        "num_params_M": r"^num_params_M:\s+([\d.]+)",
        "depth": r"^depth:\s+(\d+)",
    }
    results = {}
    for key, pattern in patterns.items():
        match = re.search(pattern, log, re.MULTILINE)
        results[key] = match.group(1) if match else None
    return results
```

### Important Implementation Notes

- **Subprocess management:** Use `subprocess.Popen` with `os.setpgrp` so the entire process group can be killed on timeout, not just the parent.
- **Temp directory:** Use `tempfile.mkdtemp()`. Clean up in a `finally` block so it's removed even on crash.
- **`pyproject.toml` for `uv run`:** `uv run train.py` needs a `pyproject.toml` in the working directory. Bundle a minimal one alongside `prepare.py` in `autoresearch/data/` with just the PyTorch dependency.
- **Log tail:** `run_log_tail` = last 100 lines of captured stdout+stderr. If the log is shorter than 100 lines, include the entire log.
- **`prepare.py` one-time setup:** The data cache at `~/.cache/autoresearch/` persists across runs. `setup()` should be called once at miner startup, not per-experiment.

### Acceptance Criteria

- [ ] `ExperimentRunner.run(valid_train_py)` returns a `RunResult` with `status="success"` and all metric fields populated when given a valid `train.py` and a GPU is available
- [ ] `ExperimentRunner.run(sleep_forever_script)` returns `status="timeout"` after 600s and kills the subprocess
- [ ] `ExperimentRunner.run(syntax_error_script)` returns `status="crash"` with `run_log_tail` containing the traceback
- [ ] Temp directories are cleaned up after every run (success, crash, and timeout)
- [ ] `setup()` runs `prepare.py` only if cache is missing; skips and logs on subsequent calls
- [ ] `parse_metrics()` correctly extracts all 9 fields from a valid summary block
- [ ] `parse_metrics()` returns `None` for missing fields when the log is truncated

### Test Expectations

- `test_parse_metrics_valid_log` — sample log string → all 9 fields parsed correctly with correct types (float/int)
- `test_parse_metrics_missing_fields` — truncated log → missing fields are `None`
- `test_parse_metrics_empty_log` — empty string → all fields `None`
- `test_runner_returns_timeout_on_slow_script` — mock a script that runs `time.sleep(9999)`, assert timeout status (use a very short timeout override for testing, not the real 600s)
- `test_runner_returns_crash_on_invalid_python` — script with `raise RuntimeError("test")` → status="crash", traceback in log tail
- `test_runner_cleans_up_temp_dir` — assert temp dir is removed after each run status
- `test_run_result_dataclass_defaults` — freshly created `RunResult()` has all None fields and status="pending"
